### PR TITLE
[nGraph] MatMul - Remove fused op and align output shape inference

### PR DESCRIPTION
**Pull request by mitruska**
To follow [MatMul spec update](https://github.com/openvinotoolkit/openvino/pull/2765) for 1D tensors this PR introduces FusedOp and decompose removal without changing current MatMul output shape inference logic (numpy/onnx aligned). 
Based on previous PR https://github.com/openvinotoolkit/openvino/pull/2212 that follows rather current spec logic.

### CODE REVIEW

File: ngraph/core/src/op/matmul.cpp [hunk 1]
```diff
@@ -113,68 +53,125 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
...
-            output_shape.insert(output_shape.end(), arg1_shape_update[arg1_rank - 1]);
+            // Get max value for all batches (max_rank - 2), COL_INDEX_DIM and ROW_INDEX_DIM are
+            // updated at the end
+            for (auto i = 0; i < max_rank - 2; i++)
+            {
+                if (low_size_matrix[i].is_dynamic() || big_size_matrix[i].is_dynamic())
+                {
+                    // Static value is assigned to output when it is > 1, otherwise
+                    // dynamic dimension is forwarded to output
+                    Dimension::merge(merged_dimension, low_size_matrix[i], big_size_matrix[i]);
```
Comment (slyalin) [inline ngraph/core/src/op/matmul.cpp:144]: Consider this valid case: both dimensions are dynamic, one has range [1..5], another one is [10..20]. Merge will return false as these ranges don't intersect. But 1 is included into the first range, so broadcasting of it is possible. Then it is a valid case and the output range should be [10..20].

Comment (slyalin) [reply ngraph/core/src/op/matmul.cpp:144]: So you need to check if 1 is included into the range, and proceed it in a special way.

File: ngraph/core/src/op/matmul.cpp [hunk 2]
```diff
@@ -113,68 +53,125 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
...
+                    if (merged_dimension.is_static() && merged_dimension.get_length() > 1)
+                    {
+                        output_shape[i] = merged_dimension;
+                    }
+                    else
+                    {
+                        output_shape[i] = Dimension::dynamic();
```
Comment (slyalin) [inline ngraph/core/src/op/matmul.cpp:151]: Please forward dynamic dimension that merge gave you above. It has narrowed range instead of pure dynamic.

File: ngraph/core/src/op/matmul.cpp [hunk 3]
```diff
@@ -113,68 +53,125 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
...
+            // Get max value for all batches (max_rank - 2), COL_INDEX_DIM and ROW_INDEX_DIM are
+            // updated at the end
+            for (auto i = 0; i < max_rank - 2; i++)
+            {
+                if (low_size_matrix[i].is_dynamic() || big_size_matrix[i].is_dynamic())
+                {
+                    // Static value is assigned to output when it is > 1, otherwise
+                    // dynamic dimension is forwarded to output
+                    Dimension::merge(merged_dimension, low_size_matrix[i], big_size_matrix[i]);
+                    if (merged_dimension.is_static() && merged_dimension.get_length() > 1)
```
Comment (slyalin) [inline ngraph/core/src/op/matmul.cpp:145]: Why does 1 have a special meaning?

File: ngraph/core/src/op/matmul.cpp (hunk 1 above)
Comment (slyalin) [reply ngraph/core/src/op/matmul.cpp:144]: @mitruska, you convinced me that handling ranges for Rank is not possible with the current nGraph code, but for each dimension separately it is still possible and we have to make it as precise as possible.

**Review by slyalin — CHANGES_REQUESTED** [review]
More attention for dynamic dimensions. Please provide relevant tests with ranges.

File: ngraph/core/src/op/matmul.cpp (hunk 3 above)
Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp]: Because 1 can be expanded to bigger dimension. 
Comment was added.

File: ngraph/core/src/op/matmul.cpp (hunk 1 above)
Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp]: Ok, I thought that we want to stay with fully dynamic dimensions. 
I've just added handling dynamic dimensions range. Please take a look.

File: ngraph/core/src/op/matmul.cpp (hunk 2 above)
Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp]: Done, please review

File: ngraph/test/type_prop/matmul.cpp [hunk 4]
```diff
@@ -108,3 +119,408 @@ TEST(type_prop, matmul_4D_transpose_b)
...
+                                              Dimension(5, 7),   // 4
+                                              Dimension(5, 7),   // 5
+                                              Dimension(1, -1),  // 6
+                                              Dimension(0, 1),   // 7
+                                              Dimension(2, 5),   // 8
+                                              Dimension(1, -1),  // 9
+                                              3,                 // 10
+                                              5,                 // 11
+                                              4};                // 12
+
```
Comment (mitruska) [inline ngraph/test/type_prop/matmul.cpp:474]: @slyalin Please confirm that dynamic dimension range tests match common expectations.

Comment (mitruska) [reply ngraph/test/type_prop/matmul.cpp:474]: Particularly `[0, 1] & [0, 1]` case, we can infer as output dim with `[0, 1]` range
or consider as static dim with single value `1` because `0` would be wrong batch dimension. 
Same way `dynamic & dynamic` which is `[0, max] & [0, max]` can be just `dynamic` or `[1, max]`.

(Init Dimension with  `-1` means Interval min/max value limit)

Comment (slyalin) [reply ngraph/test/type_prop/matmul.cpp:474]: 0 is a valid dimension value which means that the corresponding tensor is empty, but still have some dimensions. We don't have any exceptions for 0 dimensions in the specification. So we shouldn't ignore 0 by treating it as impossible.

File: ngraph/core/src/op/matmul.cpp [hunk 5]
```diff
@@ -113,68 +53,170 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
...
+            arg0_shape_tmp.erase(arg0_shape_tmp.begin() + arg0_rank - 1);
+            return PartialShape(arg0_shape_tmp);
         }
-        else
+
+        // 2D and bigger tensors cases.
+        NGRAPH_CHECK(Dimension::merge(merged_dimension,
+                                      arg0_shape_tmp[arg0_rank - 1],
+                                      arg1_shape_tmp[arg1_rank - 2]),
+                     "Incompatible matrix dimensions.");
```
Comment (slyalin) [inline ngraph/core/src/op/matmul.cpp:116]: Referring our discussion yesterday when we were not able to decide which behavior we should choose when have two dynamic dimensions that are not really compatible. It is relevant to this place and places below about incompatibilities during the broadcasting of batch dimensions. I mean cases like we trying to match [2,3] and [4,5] for example. I think it should fail (as written in this code, right?). I think we should follow this approach of aggressive failing because the original validation code for MatMul doesn't handle dynamic cases at all except prevalidate function. @mitruska, can you confirm that old MatMul::validate_and_infer_types cannot proceed with dynamic inputs?

File: ngraph/core/src/op/matmul.cpp [hunk 6]
```diff
@@ -113,68 +53,170 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
...
+                                                big_batch_matrix[i].get_min_length());
+                    if (min_dim_val > 1)
+                    {
+                        NGRAPH_CHECK(Dimension::merge(merged_dimension,
+                                                      small_batch_matrix[i],
+                                                      big_batch_matrix[i]),
+                                     "Incompatible batch dimensions bounds.");
+                        output_shape[i] = merged_dimension;
+                        continue;
+                    }
```
Comment (slyalin) [inline ngraph/core/src/op/matmul.cpp:167]: I think that you can produce shorter code by not differentiating static and dynamic cases here and below. The current code is huge and hard to follow and maintain. Dimension::merge works fine for static shapes as well. Just check if 1 is included into a range and proceed accordingly avoid explicit touching get_length attribute for a dimension. Does it make sense, @mitruska? Could you try?

**Review by slyalin — CHANGES_REQUESTED** [review]

File: ngraph/core/src/op/matmul.cpp (hunk 5 above)
Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp:116]: Old (current master) implementation for dynamic cases preserves only rank.
For the same test case as [matmul_batch_dynamic_bounds](https://github.com/openvinotoolkit/openvino/blob/9c2a38ca338fed06e75a460c696e547d121909ee/ngraph/test/type_prop/matmul.cpp#L433-L474) the output is:
> Expected equality of these values:
  matmul->get_output_partial_shape(0)
    Which is: {?,?,?,?,?,?,?,?,?,?,?,?,?}
  expected_output_shape
    Which is: {?,[10, 20],10,[5, 7],[5, 7],[5, 7],?,[0, 1],[2, 5],?,3,5,4}

File: ngraph/core/src/op/matmul.cpp (hunk 6 above)
Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp]: Thank you for the suggestion, I refactored and reduced this part of code.

File: ngraph/core/src/op/matmul.cpp [hunk 7]
```diff
@@ -113,68 +53,170 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
...
+            // i.e. arg0 shape {2, 2, 3}, arg1 shape{3}, output shape {2, 2}
+            NGRAPH_CHECK(Dimension::merge(
+                             merged_dimension, arg1_shape_tmp[0], arg0_shape_tmp[arg0_rank - 1]),
+                         "Incompatible matrix dimensions.");
+            arg0_shape_tmp.erase(arg0_shape_tmp.begin() + arg0_rank - 1);
+            return PartialShape(arg0_shape_tmp);
         }
-        else
+
+        // 2D and bigger tensors cases.
```
[the remark this answers is not in the document]
Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp]: Done

File: ngraph/core/src/op/matmul.cpp [hunk 8]
```diff
@@ -113,68 +53,170 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
...
-            output_shape = arg0_shape_update;
-        }
-        else if (arg0_rank == 2 && arg1_rank == 2)
-        {
-            NGRAPH_CHECK(arg0_shape_update[1] == arg1_shape_update[0], "Incompatible arg shapes");
-            output_shape = Shape{arg0_shape_update[0], arg1_shape_update[1]};
+            // i.e. arg0 shape {2, 2, 3}, arg1 shape{3}, output shape {2, 2}
+            NGRAPH_CHECK(Dimension::merge(
+                             merged_dimension, arg1_shape_tmp[0], arg0_shape_tmp[arg0_rank - 1]),
+                         "Incompatible matrix dimensions.");
```
[the remark this answers is not in the document]
Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp]: Done

File: ngraph/core/src/op/matmul.cpp [hunk 9]
```diff
@@ -113,68 +53,170 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
...
-            // i.e., arg0 shape {3}, arg1 shape{2, 3, 2}, output shape {2, 2}
-            NGRAPH_CHECK(arg0_shape_update[0] == arg1_shape_update[arg1_rank - 2],
-                         "Incompatible arg shapes");
-            arg1_shape_update.erase(arg1_shape_update.begin() + arg1_rank - 2);
-            output_shape = arg1_shape_update;
+            // i.e. arg0 shape {3}, arg1 shape{2, 3, 2}, output shape {2, 2}
+            NGRAPH_CHECK(Dimension::merge(merged_dimension,
+                                          arg0_shape_tmp[0],
+                                          arg1_shape_tmp[arg1_shape_tmp.size() - 2]),
+                         "Incompatible matrix dimensions.");
```
[the remark this answers is not in the document]
Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp]: Done

File: ngraph/core/src/op/matmul.cpp [hunk 10]
```diff
@@ -113,68 +53,170 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
...
-        if (transpose_a && arg0_rank > 1)
+        // Temporary Dimension vectors to calculate output shape
+        std::vector<Dimension> arg0_shape_tmp(arg0_shape);
+        std::vector<Dimension> arg1_shape_tmp(arg1_shape);
+
+        // Result of merging compatible dimensions for validation usage
+        auto merged_dimension = Dimension::dynamic();
+
+        // Transpose attributes are ignored for 1D tensors.
+        if (arg0_rank == 1 && arg1_rank == 1)
```
[the remark this answers is not in the document]
Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp]: Done

**Review by tomdol — APPROVED** [review]

File: ngraph/core/src/op/matmul.cpp (hunk 5 above)
Comment (slyalin) [reply ngraph/core/src/op/matmul.cpp:107]: So it doesn't catch case when dimensions are not compatible and then just set ? for corresponding output dimension?

Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp:107]: Yes, unfortunately current master doesn't throw with incompatible batch dynamic bounds like: https://github.com/openvinotoolkit/openvino/blob/2c7360938ae35291c76f5bb2ad4d17d3c514579c/ngraph/test/type_prop/matmul.cpp#L552-L571

Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp:107]: So I can check here https://github.com/openvinotoolkit/openvino/blob/2c7360938ae35291c76f5bb2ad4d17d3c514579c/ngraph/core/src/op/matmul.cpp#L135-L139 returned value from `Dimension::merge` and throw only for static, otherwise set `Dimensions::dynamic()`

Comment (slyalin) [reply ngraph/core/src/op/matmul.cpp:107]: Looks like we have to implement the same logic for another Dimension::merge call below for other dimensions. I would propose to introduce a helper function for this backward compatible goal which should be used instead of raw Dimension::merge.

**Comment by ilyachur** [comment]
@mitruska I think git merge was incorrect. Please revert the merge and remove unnecessary files.

**Comment by mitruska** [comment]
@ilyachur You're right, done.

**Review by mbencer — APPROVED** [review]

File: ngraph/core/src/op/matmul.cpp [hunk 11]
```diff
@@ -113,68 +53,152 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
...
+            // For example {S} will be reshaped to {S, 1}.
+            arg1_shape_tmp.insert(arg1_shape_tmp.end(), 1);
+            arg1_rank = arg1_shape_tmp.size();
         }
-        else if (arg1_rank == 1)
+
+        // Check matrices dimensions compatibility,
+        // COL_INDEX_DIM of the first matrix has to match ROW_INDEX_DIM of the second matrix.
+        auto merged_dimension = Dimension::dynamic();
+        auto arg0_col_dim = arg0_shape_tmp[arg0_rank - 1];
```
Comment (ilyachur) [inline ngraph/core/src/op/matmul.cpp:104]: What happens if rank is dynamic?

Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp:104]: This validation is for static rank. For dynamic rank, `PartialShape::dynamic()` will be returned. https://github.com/openvinotoolkit/openvino/blob/e0d0ee561b9c0dc6668f08112d2e8924618d47f7/ngraph/core/src/op/matmul.cpp#L284-L299

**Review by postrational — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

File: ngraph/core/src/op/matmul.cpp [hunk 12]
```diff
@@ -113,68 +53,152 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
...
+                     (arg1_rank - 2));
+
+        // 3. If ranks of input arguments are different after steps 1 and 2,
+        // the smaller tensor is unsqueezed from the left side of the shape
+        // by necessary number of axes to make both shapes of the same rank.
+        std::vector<Dimension> small_batch_matrix =
+            arg0_rank > arg1_rank ? arg1_shape_tmp : arg0_shape_tmp;
+        std::vector<Dimension> big_batch_matrix =
+            arg0_rank > arg1_rank ? arg0_shape_tmp : arg1_shape_tmp;
+        std::vector<Dimension> output_shape(big_batch_matrix);
```
Comment (slyalin) [inline ngraph/core/src/op/matmul.cpp:125]: Please move this block under the next condition and modify it because these manipulations are required only when arg0_rank != arg1_rank. Introducing new small/big batch matrix variables look to be a waste of time. Insert 1's directly to one of arg0_shape_tmp or arg1_shape_tmp depending on the ranks comparison. Define output_shape after the condition when both vectors are of the same size.

File: ngraph/core/src/op/matmul.cpp [hunk 13]
```diff
@@ -113,68 +53,152 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
...
-        else if (arg0_rank == 2 && arg1_rank == 2)
+
+        // 4. Usual rules of the broadcasting are applied for batch dimensions.
+        // Broadcast all batches (last two dimensions represent matrix),
+        // expand dim with value 1 to bigger dim if dimensions are not equal.
+        for (auto i = 0; i < big_batch_matrix.size() - 2; i++)
         {
-            NGRAPH_CHECK(arg0_shape_update[1] == arg1_shape_update[0], "Incompatible arg shapes");
-            output_shape = Shape{arg0_shape_update[0], arg1_shape_update[1]};
+            // Dimension with value 1 can be expanded to any bigger
```
Comment (slyalin) [inline ngraph/core/src/op/matmul.cpp:139]: This comment is not so relevant to the next code without additional explanations. It is more suitable for case `if (lower_bound <= 1)`.

File: ngraph/core/src/op/matmul.cpp [hunk 14]
```diff
@@ -113,68 +53,152 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
...
+                Dimension::value_type upper_bound, lower_bound;
+                lower_bound = std::max(small_batch_matrix[i].get_min_length(),
+                                       big_batch_matrix[i].get_min_length());
+
+                if (lower_bound <= 1)
+                {
+                    // Both of the dimensions have 1 in range,
+                    // upper_bound is the maximum of the each range highest possible value.
+                    upper_bound = std::max(small_batch_matrix[i].get_interval().get_max_val(),
+                                           big_batch_matrix[i].get_interval().get_max_val());
```
Comment (slyalin) [inline ngraph/core/src/op/matmul.cpp:170]: Use `get_max_length`.

File: ngraph/core/src/op/matmul.cpp [hunk 15]
```diff
@@ -113,68 +53,152 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
...
+        for (auto i = 0; i < big_batch_matrix.size() - 2; i++)
         {
-            NGRAPH_CHECK(arg0_shape_update[1] == arg1_shape_update[0], "Incompatible arg shapes");
-            output_shape = Shape{arg0_shape_update[0], arg1_shape_update[1]};
+            // Dimension with value 1 can be expanded to any bigger
+            auto min_dim_val = std::min(small_batch_matrix[i].get_min_length(),
+                                        big_batch_matrix[i].get_min_length());
+            if (min_dim_val > 1)
+            {
+                auto merged_dimension = Dimension::dynamic();
```
Comment (slyalin) [inline ngraph/core/src/op/matmul.cpp:144]: Add comment that this is for case when both dimensions don't have 1 in range, which means that usual merge is enough.

File: ngraph/core/src/op/matmul.cpp [hunk 16]
```diff
@@ -113,68 +53,152 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
...
+                             " with dimension=",
+                             big_batch_matrix[i],
+                             " at index=",
+                             i);
+
+                output_shape[i] = merged_dimension;
+            }
+            else
+            {
+                Dimension::value_type upper_bound, lower_bound;
```
Comment (slyalin) [inline ngraph/core/src/op/matmul.cpp:161]: Here is where that comment about 1 is relevant. Please add more details to make the code more readable.

**Review by slyalin — CHANGES_REQUESTED** [review]

File: ngraph/core/src/op/matmul.cpp (hunk 14 above)
Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp:170]: [Dimension::get_max_length()](https://github.com/openvinotoolkit/openvino/blob/9c509e5f416cce29dddd3bd1de292055df3765c5/ngraph/core/src/dimension.cpp#L140) will return `-1` for the upper bound limit `Interval::s_max`. 
So `std::max` for `-1` as upper bound limit won't be correct.
I use `dimension.get_interval().get_max_val()` to get `Interval::s_max` value instead of `-1`.

Comment (slyalin) [reply ngraph/core/src/op/matmul.cpp:170]: You are right. This is an artifact (really an issue) of PartialShape that I've missed.

File: ngraph/core/src/op/matmul.cpp (hunk 12 above)
Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp]: Done, I think this part looks much better now. Thanks.

File: ngraph/core/src/op/matmul.cpp (hunk 13 above)
Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp]: Done

File: ngraph/core/src/op/matmul.cpp (hunk 15 above)
Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp:150]: Done

File: ngraph/core/src/op/matmul.cpp (hunk 16 above)
Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp:167]: Done

Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp]: Done

File: ngraph/core/src/op/matmul.cpp [hunk 17]
```diff
@@ -113,68 +53,162 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
...
-            arg0_shape_update.erase(arg0_shape_update.begin() + arg0_rank - 1);
-            output_shape = arg0_shape_update;
+            // Expand small_batch_matrix (with 1) to have the same rank as big_batch_matrix
+            // size_t delta_rank = big_batch_matrix.size() - small_batch_matrix.size();
+            // small_batch_matrix.insert(small_batch_matrix.begin(), delta_rank, 1);
+            int64_t delta_rank = arg0_rank - arg1_rank;
+            delta_rank < 0 ? arg0_shape_tmp.insert(arg0_shape_tmp.begin(), std::abs(delta_rank), 1)
+                           : arg1_shape_tmp.insert(arg1_shape_tmp.begin(), std::abs(delta_rank), 1);
+            arg0_rank = arg0_shape_tmp.size();
+            arg1_rank = arg1_shape_tmp.size();
```
Comment (slyalin) [inline ngraph/core/src/op/matmul.cpp:132]: arg1_rank is never used below. Ternary operator is used instead of if/else statement (why?). Comment is not relevant to the code. OK, let me just propose this short form without all those `abs` etc.:

```
if (arg0_rank < arg1_rank)
    arg0_shape_tmp.insert(arg0_shape_tmp.begin(), arg1_rank - arg0_rank, 1);
else if (arg0_rank > arg1_rank)
    arg1_shape_tmp.insert(arg1_shape_tmp.begin(), arg0_rank - arg1_rank, 1);
auto max_rank = arg0_shape_tmp.size();   // both arg0_shape_tmp and arg1_shape_tmp have identical size now
```

**Review by slyalin — CHANGES_REQUESTED** [review]

**Review by slyalin — CHANGES_REQUESTED** [review]
Please improve one last part that I've pointed in the comment.

File: ngraph/core/src/op/matmul.cpp (hunk 17 above)
Comment (mitruska) [reply ngraph/core/src/op/matmul.cpp]: Thanks, done

**Review by slyalin — APPROVED** [review]
